src/transport/CertificateAuthenticatedSessionEstablishment.cpp

```
@@ -0,0 +1,415 @@
+
+    msgBuf->SetStart(msgBuf->Start() - headerSize);
+    err = payloadHeader.Encode(msgBuf->Start(), msgBuf->DataLength(), &actualEncodedHeaderSize);
+    SuccessOrExit(err);
+    VerifyOrExit(headerSize == actualEncodedHeaderSize, err = CHIP_ERROR_INTERNAL);
+
+    err = mDelegate->SendPairingMessage(PacketHeader()
+                                            .SetSourceNodeId(mLocalNodeId)
+                                            .SetDestinationNodeId(mConnectionState.GetPeerNodeId())
+                                            .SetEncryptionKeyID(mConnectionState.GetLocalKeyID()),
+                                        mConnectionState.GetPeerAddress(), std::move(msgBuf));
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR CertificateAuthenticatedSessionEstablishment::EstablishSession(const Transport::PeerAddress peerAddress, NodeId myNodeId,
+                                                                          NodeId peerNodeId, uint16_t myKeyId,
+                                                                          AuthenticatedSessionEstablishmentDelegate * delegate)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    mConnectionState.SetPeerAddress(peerAddress);
+    mConnectionState.SetPeerNodeId(peerNodeId);
+
+    err = SendSigmaR1();
+    SuccessOrExit(err);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        Clear();
+    }
+    return err;
+}
+
+CHIP_ERROR CertificateAuthenticatedSessionEstablishment::DeriveSecureSession(const uint8_t * info, size_t info_len,
+                                                                             SecureSession & session)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrExit(info != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(info_len > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(mPairingComplete, err = CHIP_ERROR_INCORRECT_STATE);
+
+    // ToDo: Use asymmetric keypair to create the secure session
+    // err = session.Init();
```

For now, shouldn't we return an error here?

---

We don't have the actual parameters to pass to the Init() function. I left it as a stubbed out code just as a reference on which function to call.

---

Right, but I mean we should not succeed on `DeriveSecureSession` if it doesn't work, right?

---

ok got it. Yeah, we can just return error for time being, since it's not implemented yet.